lib/network/level-net/sio-stream.js

```diff
@@ -0,0 +1,372 @@
+        super({ objectMode: true });
+        this._initOutputStream(socket, streamId, maxPendingAck,
+                               ackTimeoutMs);
+    }
+
+    _initOutputStream(socket, streamId, maxPendingAck, ackTimeoutMs) {
+        this.socket = socket;
+        this.streamId = streamId;
+        this.on('finish', () => {
+            debug('output stream finished', this.streamId);
+            this.socket._finish(this.streamId, err => {
+                debug('ack finish', this.streamId, 'err', err);
+            });
+        });
+        this.on('error', err => {
+            debug('output stream error', this.streamId);
+            // notify remote of the error
+            this.socket._error(this.streamId, err);
+        });
+
+        // This is used for queuing flow control, don't issue more
+        // than maxPendingAck requests (events) that have not been
+        // acked yet
+        this.maxPendingAck = maxPendingAck;
+        this.ackTimeoutMs = ackTimeoutMs;
+
+        this.nPendingAck = 0;
+    }
+
+    _write(chunk, encoding, callback) {
+        return this._writev([{ chunk }], callback);
+    }
+
+    _writev(chunks, callback) {
+        const payload = chunks.map(chunk => chunk.chunk);
+
+        debug(`_writev(${JSON.stringify(payload)}, ...)`);
+        this.nPendingAck += 1;
+        const timeoutInfo =
+            `stream timeout: did not receive ack after ${this.ackTimeoutMs}ms`;
+        async.timeout(cb => {
+            this.socket._write(this.streamId, payload, cb);
+        }, this.ackTimeoutMs, timeoutInfo)(
+            err => {
+                debug(`ack stream-data ${this.streamId}
+                      (${JSON.stringify(payload)}):`, err);
+                if (this.nPendingAck === this.maxPendingAck) {
+                    callback();
+                }
+                this.nPendingAck -= 1;
+                if (err) {
+                    // notify remote of the error (timeout notably)
+                    debug('stream error:', err);
+                    this.socket._error(this.streamId, err);
+                    // stop the producer
+                    this.socket.destroyStream(this.streamId);
+                }
+            });
+        if (this.nPendingAck < this.maxPendingAck) {
+            callback();
+        }
+    }
+}
+
+class SIOInputStream extends stream.Readable {
+    constructor(socket, streamId) {
+        super({ objectMode: true });
+        this.socket = socket;
+        this.streamId = streamId;
+        this._readState = {
+            pushBuffer: [],
+            readable: false,
+        };
+        this.on('end', () => {
+            debug('input stream end', this.streamId);
+        });
+    }
+
+    destroy() {
+        debug('destroy called', this.streamId);
+        this._destroyed = true;
+        this.pause();
+        this.removeAllListeners('data');
+        this.removeAllListeners('end');
+        this._readState = {
+            pushBuffer: [],
```

Comment:
What do you mean more precisely, .e.g in case we keep a reference to _readState? Normally not it should be safe (and normally having an empty pushBuffer is not essential because it should not be used anymore at this stage, but it's cleaner to clean it I think)